django/utils/choices.py

```diff
@@ -73,36 +73,41 @@ def normalize_choices(value, *, depth=0):
     # Avoid circular import when importing django.forms.
     from django.db.models.enums import ChoicesType
 
-    match value:
-        case BaseChoiceIterator() | Promise() | bytes() | str():
-            # Avoid prematurely normalizing iterators that should be lazy.
-            # Because string-like types are iterable, return early to avoid
-            # iterating over them in the guard for the Iterable case below.
-            return value
-        case ChoicesType():
-            # Choices enumeration helpers already output in canonical form.
-            return value.choices
-        case Mapping() if depth < 2:
-            value = value.items()
-        case Iterator() if depth < 2:
-            # Although Iterator would be handled by the Iterable case below,
-            # the iterator would be consumed prematurely while checking that
-            # its elements are not string-like in the guard, so we handle it
-            # separately.
-            pass
-        case Iterable() if depth < 2 and not any(
-            isinstance(x, (Promise, bytes, str)) for x in value
-        ):
-            # String-like types are iterable, so the guard above ensures that
-            # they're handled by the default case below.
-            pass
-        case Callable() if depth == 0:
+    if isinstance(value, (BaseChoiceIterator, Promise, bytes, str)):
```

Comment:
I would wait for the PyPy Team before implementing any temporary solution here.

Reply:
Sure. Being familiar with the code, it was straightforward.

Was helpful to put this in for now to help see the wood for the trees...

Reply:
This is in a separate commit, so we can remove it easily if PyPy gets a fix.

Reply:
We can remove it now as it's fixed in PyPY ([logs](https://github.com/django/django/actions/runs/6976283957/job/18985932640)).

Reply:
I've confirmed the fix with a local nightly build too. Have dropped the commit.